### PR TITLE
fix(TabBar): Don't call SelectedIndex setter with the same old value to let it work with bindings

### LINKED ISSUE #562: [Windows][TabBar] SelectedIndex doesn't update selected item visual state
## Current behavior

On Windows, we need to set `SelectedIndex` as a `TwoWay` binding if we want the selected item visual state to update.

Animation_90

## Expected behavior

On Windows, we should not need to set `SelectedIndex` as a `TwoWay` binding if we want the selected item visual state to update.

Please note that I know that if we don't set  `SelectedIndex` as a `TwoWay` binding the selected index property won't be updated since it's an expected behavior.

Animation_91

## How to reproduce it (as minimally and precisely as possible)

1. Download [TabBarSample.zip](https://github.com/unoplatform/uno.toolkit.ui/files/11401095/TabBarSample.zip)
2. Start on UWP (Would also be able to reproduce on WinUI)
3. Click on the buttons
4. Notice that the selected index is not updating the item visual state.

Nuget Package: Uno.Toolkit.UI

Package Version(s): 2.6.0-dev.40

Affected platform(s):

- [ ] WebAssembly
- [ ] Android
- [ ] iOS
- [ ] macOS (AppKit)
- [ ] Mac Catalyst
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Linux Framebuffer
  - [ ] Tizen
- [x] Windows

IDE:

- [x] Visual Studio 2022
- [ ] Visual Studio 2019
- [ ] Visual Studio Code
- [ ] Visual Studio for Mac
- [ ] Rider Windows
- [ ] Rider macOS

Relevant plugins:

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): Closes #562

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] [Docs](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc) have been added/updated
- [ ] [Runtime Tests and/or UI Tests](https://platform.uno/docs/articles/contributing/guidelines/creating-tests.html) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal)
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by Youssef1313** [comment]
@kazo0 @Xiaoy312 Does this sound reasonable?

File: src/Uno.Toolkit.UI/Controls/TabBar/TabBar.cs
Comment (jeromelaban) [inline src/Uno.Toolkit.UI/Controls/TabBar/TabBar.cs:1]: @kazo0 do we have tabbar runtime or UI tests?

Comment (kazo0) [reply src/Uno.Toolkit.UI/Controls/TabBar/TabBar.cs:1]: yes ;) 

@Youssef1313 let's get some runtime tests for this

File: src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs [hunk 1]
```diff
@@ -388,5 +391,40 @@ public async Task Verify_Indicator_Placement(IndicatorPlacement placement)
...
+			var SUT = new TabBar();
+			SUT.Items.Add(item1);
+			SUT.Items.Add(item2);
+			SUT.DataContext = new MyViewModel();
+			SUT.SetBinding(TabBar.SelectedIndexProperty, new Binding() { Mode = BindingMode.OneWay, Path = new PropertyPath("P") });
+
+			await UnitTestUIContentHelperEx.SetContentAndWait(SUT);
+			Assert.IsNotNull(SUT.GetBindingExpression(TabBar.SelectedIndexProperty));
+			SUT.SelectedIndex = 0;
+			// Possible Uno bug? Explicit setting of SelectedIndex should have cleared the binding.
```
Comment (kazo0) [inline src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs:410]: Should we log an issue for this? Does it work on Windows?

**Review by kazo0 — APPROVED** [review]

File: src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs (hunk 1 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs:410]: setting value (almost?) never clears the binding; in fact, that how you get value propagated through the binding(s)
if you want to clear the bindings, call `SUT.ClearValue(TabBar.SelectedIndexProperty);`

File: src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs [hunk 2]
```diff
@@ -388,5 +391,40 @@ public async Task Verify_Indicator_Placement(IndicatorPlacement placement)
...
+			//Assert.IsNull(SUT.GetBindingExpression(TabBar.SelectedIndexProperty));
+		}
+
+		private class MyViewModel : INotifyPropertyChanged
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs:414]: use a class name that is relevant to the associated test

File: src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs [hunk 3]
```diff
@@ -388,5 +391,40 @@ public async Task Verify_Indicator_Placement(IndicatorPlacement placement)
...
+		public async Task Verify_SelectedIndex_Not_Set_Unnecessarily()
+		{
+			var item1 = new TabBarItem();
+			var item2 = new TabBarItem();
+
+			var SUT = new TabBar();
+			SUT.Items.Add(item1);
+			SUT.Items.Add(item2);
+			SUT.DataContext = new MyViewModel();
+			SUT.SetBinding(TabBar.SelectedIndexProperty, new Binding() { Mode = BindingMode.OneWay, Path = new PropertyPath("P") });
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs:405]:
```suggestion
			SUT.SetBinding(TabBar.SelectedIndexProperty, new Binding(nameof(MyViewModel.P)) { Mode = BindingMode.OneWay });
```

File: src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs:410]: @Xiaoy312 This is what I noticed on WinUI. It gets cleared for OneWay, but not TwoWay.

Comment (Youssef1313) [reply src/Uno.Toolkit.RuntimeTests/Tests/TabBarTests.cs:410]: My test on Windows was having a `<TextBox x:Name="tb" Text="{x:Bind Something, Mode=OneWay}" />`, then get the binding expression and find it's non-null, then set `tb.Text = something`, then get the binding expression again and find it's null

**Pull request merged**